File: 1/2020.10.10/74.Search a 2D Matrix/python/solution.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def binary_search(start, end):
            if start > end:
                return False
            mid = start + (end - start) // 2
            row, col = mid // n, mid % n
            if target == matrix[row][col]:
                return True
            elif target < matrix[row][col]:
                return binary_search(start, mid-1)
            else:
                return binary_search(mid+1, end)
        if len(matrix) == 0:
            return False
        m, n = len(matrix), len(matrix[0])
        return binary_search(0, m*n-1)

File: 1/2020.10.10/74.Search a 2D Matrix/java/solution.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix.length == 0) {
            return false;
        }
        int m = matrix.length, n = matrix[0].length;
        int left = 0, right = m * n - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            int pivot = matrix[mid/n][mid%n];
            if (target == pivot) {
                return true;
            } else if (target < pivot) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return false;
    }
}
